pkg/util/wait/backoff.go

```diff
@@ -0,0 +1,73 @@
+package wait
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+)
+
+// UntilWithBackoff runs f in a loop until context indicates finished. It
+// applies backoff depending on the SpeedSignal f returns.  Backoff increases
+// exponentially, ranging from 1ms to 100ms.
+func UntilWithBackoff(ctx context.Context, f func(context.Context) SpeedSignal) {
+	mgr := defaultBackoffManager()
+	wait.BackoffUntil(func() {
+		mgr.toggleBackoff(f(ctx))
+	}, mgr, false, ctx.Done())
+}
+
+// SpeedSignal indicates whether we should run the function again immediately,
+// or apply backoff.
+type SpeedSignal interface{}
+
+// KeepGoing signals to continue immediately.
+type KeepGoing struct{ SpeedSignal }
+
+// SlowDown signals to backoff.
+type SlowDown struct{ SpeedSignal }
+
+func (s *speedyBackoffManager) toggleBackoff(speedSignal SpeedSignal) {
+	switch speedSignal.(type) {
+	case KeepGoing:
+		if *s.backingOff {
+			*s.backingOff = false
+			s.rateLimiter.Forget("")
+		}
+	case SlowDown:
+		*s.backingOff = true
+	}
+}
+
+type speedyBackoffManager struct {
+	backingOff  *bool
+	rateLimiter workqueue.RateLimiter
+	timer       clock.Timer
+	wait.BackoffManager
+}
+
+func (s speedyBackoffManager) Backoff() clock.Timer {
+	s.timer.Reset(s.backoffDuration())
+	return s.timer
+}
+
+func defaultBackoffManager() speedyBackoffManager {
+	// create and drain timer, allowing reuse of same timer via timer.Reset
+	timer := clock.RealClock{}.NewTimer(0)
```

Comment:
```suggestion
func defaultBackoffManager(c clock.Clock) speedyBackoffManager {
	// create and drain timer, allowing reuse of same timer via timer.Reset
	timer := c.NewTimer(0)
```

I'm wondering if we should receive the `clock` so that we can pass the fake clock in the scheduler unit tests.

Reply:
note: the scheduler unit tests don't end up using `UntilWithBackoff`, as only `schedule` is tested, rather than `Start` (where `UntilWithBackoff` is called). if we end up adding a unit test which uses `UntilWithBackoff`, I can make this change :)